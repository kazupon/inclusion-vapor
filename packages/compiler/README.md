# svelte-vapor-compiler

> Vapor compiler for svelte

## ©️ License

[MIT](http://opensource.org/licenses/MIT)
