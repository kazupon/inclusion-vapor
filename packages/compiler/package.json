{
  "name": "svelte-vapor-compiler",
  "description": "Vapor compiler for svelte",
  "private": true,
  "version": "0.0.0",
  "author": {
    "name": "kazuya kawaguchi",
    "email": "kawakazu80@gmail.com"
  },
  "license": "MIT",
  "funding": "https://github.com/sponsors/kazupon",
  "bugs": {
    "url": "https://github.com/kazupon/svelte-vapor/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/kazupon/svelte-vapor.git",
    "directory": "packages/compiler"
  },
  "keywords": [
    "svelte",
    "vapor",
    "compiler"
  ],
  "homepage": "https://github.com/kazupon/svelte-vapor#readme",
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">= 18.18"
  },
  "type": "module",
  "files": [
    "dist"
  ],
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    }
  },
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "tsdown"
  },
  "dependencies": {
    "@babel/parser": "^7.24.8",
    "@vue-vapor/compiler-dom": "latest",
    "@vue-vapor/compiler-vapor": "latest",
    "@vue-vapor/shared": "latest",
    "svelte": "4.2.18"
  },
  "devDependencies": {
    "@babel/types": "^7.24.9",
    "tsdown": "^0.2.3"
  }
}
