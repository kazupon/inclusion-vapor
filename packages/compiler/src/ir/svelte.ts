// SPDX-License-Identifier: MIT
// Author: kazuya kawaguchi (a.k.a. kazupon)

import { makeMap, isObject } from '@vue-vapor/shared'

import type { SourceLocation } from '@vue-vapor/compiler-dom'
import type {
  Element as SvelteElement,
  BaseNode as SvelteBaseNode,
  Comment as SveletComment,
  Text as SvelteText,
  MustacheTag as SvelteMustacheTag
} from 'svelte/types/compiler/interfaces'

export const isBuiltInDirective: ReturnType<typeof makeMap> = /*#__PURE__*/ makeMap(
  // TODO: add svelte built-in directives
  ''
  //'bind,cloak,else-if,else,for,html,if,model,on,once,pre,show,slot,text,memo',
)

const SVELTE_ELEMENT_TYPES = new Set([
  'InlineComponent',
  'SlotTemplate',
  'Title',
  'Slot',
  'Element',
  'Head',
  'Options',
  'Window',
  'Document',
  'Body'
])

export function isSvelteElement(node: unknown): node is SvelteElement {
  return isObject(node) && 'type' in node && SVELTE_ELEMENT_TYPES.has((node as SvelteBaseNode).type)
}

const SVELETE_MUSTACHE_TAG_TYPES = new Set(['MustacheTag', 'RawMustacheTag'])

export function isSvelteMustacheTag(node: unknown): node is SvelteMustacheTag {
  return (
    isObject(node) &&
    'type' in node &&
    SVELETE_MUSTACHE_TAG_TYPES.has((node as SvelteBaseNode).type)
  )
}

export function isSvelteText(node: unknown): node is SvelteText {
  return isObject(node) && 'type' in node && node.type === 'Text'
}

export function isSvelteComment(node: unknown): node is SveletComment {
  return isObject(node) && 'type' in node && node.type === 'Comment'
}

export function convertToSourceLocation(node: SvelteBaseNode, source: string): SourceLocation {
  return {
    start: {
      offset: node.start,
      line: -1,
      column: -1
    },
    end: {
      offset: node.end,
      line: -1,
      column: -1
    },
    source
  }
}

export type {
  Ast as SvelteAst,
  Element as SvelteElement,
  TemplateNode as SvelteTemplateNode,
  Text as SvelteText,
  MustacheTag as SvelteMustacheTag,
  Comment as SvelteComment,
  Attribute as SvelteAttribute,
  SpreadAttribute as SvelteSpreadAttribute,
  BaseDirective as SvelteBaseDirective,
  BaseNode as SvelteBaseNode
} from 'svelte/types/compiler/interfaces'
