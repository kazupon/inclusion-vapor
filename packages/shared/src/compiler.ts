// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)

export * from './ast.ts'
export * from './htmlNesting.ts'
export * from './types.ts'
