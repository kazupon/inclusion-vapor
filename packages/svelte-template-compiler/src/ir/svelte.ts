// SPDX-License-Identifier: MIT
// Author: kazuya kawaguchi (a.k.a. kazupon)

import { isObject, makeMap } from '@vue-vapor/shared'

import type { SourceLocation } from '@vue-vapor/compiler-dom'
import type {
  Comment as SveletComment,
  Attribute as SvelteAttribute,
  BaseExpressionDirective as SvelteBaseExpressionDirective,
  BaseNode as SvelteBaseNode,
  ComponentTag as SvelteComponentTag,
  Directive as SvelteDirective,
  Element as SvelteElement,
  ElseBlock as SvelteElseBlock,
  IfBlock as SvelteIfBlock,
  MustacheTag as SvelteMustacheTag,
  ShorthandAttribute as SvelteShorthandAttribute,
  SpreadAttribute as SvelteSpreadAttribute,
  StyleDirective as SvelteStyleDirective,
  Text as SvelteText
} from 'svelte/types/compiler/interfaces'
import type { SvelteTemplateNode } from './svelte'

export const isBuiltInDirective: ReturnType<typeof makeMap> = /*#__PURE__*/ makeMap(
  // TODO: add svelte built-in directives
  ''
  //'bind,cloak,else-if,else,for,html,if,model,on,once,pre,show,slot,text,memo',
)

const SVELTE_ELEMENT_TYPES = new Set([
  'InlineComponent',
  'SlotTemplate',
  'Title',
  'Slot',
  'Element',
  'Head',
  'Options',
  'Window',
  'Document',
  'Body'
])

export function isSvelteElement(node: unknown): node is SvelteElement {
  return isObject(node) && 'type' in node && SVELTE_ELEMENT_TYPES.has((node as SvelteBaseNode).type)
}

const SVELTE_DIRECTIVE_TYPES = new Set([
  'Action',
  'Animation',
  'Binding',
  'Class',
  'StyleDirective',
  'EventHandler',
  'Let',
  'Ref',
  'Transition'
])

export function isSvelteDirective(node: unknown): node is SvelteDirective {
  return (
    isObject(node) && 'type' in node && SVELTE_DIRECTIVE_TYPES.has((node as SvelteBaseNode).type)
  )
}

export function isSvelteMustacheTag(node: unknown): node is SvelteMustacheTag {
  return isObject(node) && 'type' in node && node.type === 'MustacheTag'
}

export function isSvelteText(node: unknown): node is SvelteText {
  return isObject(node) && 'type' in node && node.type === 'Text'
}

export function isSvelteComponentTag(node: unknown): node is SvelteComponentTag {
  return (
    isObject(node) &&
    'type' in node &&
    node.type === 'InlineComponent' &&
    'name' in node &&
    node.name === 'svelte:component'
  )
}

export function isSvelteComment(node: unknown): node is SveletComment {
  return isObject(node) && 'type' in node && node.type === 'Comment'
}

export function isSvelteAttribute(node: unknown): node is SvelteAttribute {
  return isObject(node) && 'type' in node && node.type === 'Attribute'
}

export function isSvelteSpreadAttribute(node: unknown): node is SvelteSpreadAttribute {
  return isObject(node) && 'type' in node && node.type === 'Spread'
}

export function isSvelteShorthandAttribute(node: unknown): node is SvelteShorthandAttribute {
  return isObject(node) && 'type' in node && node.type === 'AttributeShorthand'
}

export function isSvelteEventHandler(node: unknown): node is SvelteBaseExpressionDirective {
  return isObject(node) && 'type' in node && node.type === 'EventHandler'
}

export function isSvelteElseBlock(node: unknown): node is SvelteElseBlock {
  return isObject(node) && 'type' in node && node.type === 'ElseBlock'
}

export function isIfBlockOnTop(node: SvelteIfBlock): boolean {
  return node.type === 'IfBlock' && !node.elseif
}

export function isIfBlockOnElseBlock(node: SvelteIfBlock): boolean {
  return node.type === 'IfBlock' && !!node.elseif
}

export function isSvelteBindingDirective(node: unknown): node is SvelteBaseExpressionDirective {
  return isObject(node) && 'type' in node && node.type === 'Binding'
}

export function isSvelteClassDirective(node: unknown): node is SvelteBaseExpressionDirective {
  return isObject(node) && 'type' in node && node.type === 'Class'
}

export function isSvelteStyleDirective(node: unknown): node is SvelteStyleDirective {
  return isObject(node) && 'type' in node && node.type === 'StyleDirective'
}

// TODO: more refactor the above `isSvelteBindingDirective`
export function isSvelteLetDirective(node: unknown): node is SvelteBaseExpressionDirective {
  return isObject(node) && 'type' in node && node.type === 'Let'
}

export function findAttrs(node: SvelteTemplateNode, name: string): SvelteAttribute | undefined {
  const attrs = node.attributes || [] // eslint-disable-line @typescript-eslint/no-unsafe-assignment
  for (const attr of attrs) {
    if (isSvelteAttribute(attr) && attr.name === name) {
      return attr
    }
  }
  return
}

export interface SvelteCompileError {
  code: string
  start: {
    line: number
    column: number
    character: number
  }
  end: {
    line: number
    column: number
    character: number
  }
  pos: number
  filename: string
  frame: string
}

export function isSvelteParseError(error: unknown): error is SvelteCompileError {
  return (
    isObject(error) &&
    error.constructor.name === 'CompileError' &&
    'code' in error &&
    'start' in error &&
    'end' in error &&
    'pos' in error &&
    'frame' in error
  )
}

export type CompatLocationable = {
  start: number | { line: number; column: number; offset?: number }
  end: number | { line: number; column: number; offset?: number }
  source?: string
}

// TODO: We need to extend svelte AST for location
export function convertToSourceLocation(node: CompatLocationable, source: string): SourceLocation {
  const loc = {
    start: {
      offset: -1,
      line: -1,
      column: -1
    },
    end: {
      offset: -1,
      line: -1,
      column: -1
    },
    source
  }

  function normalize(node: CompatLocationable, position: 'start' | 'end'): void {
    if (isObject(node[position])) {
      loc[position].line = node[position].line
      loc[position].column = node[position].column
      if (node[position].offset) {
        loc[position].offset = node[position].offset
      }
    } else {
      loc[position].offset = node[position]
    }
  }

  normalize(node, 'start')
  normalize(node, 'end')

  if (node.source) {
    loc.source = node.source
  }

  return loc
}

export type {
  Ast as SvelteAst,
  Attribute as SvelteAttribute,
  BaseDirective as SvelteBaseDirective,
  BaseExpressionDirective as SvelteBaseExpressionDirective,
  BaseNode as SvelteBaseNode,
  Comment as SvelteComment,
  ComponentTag as SvelteComponentTag,
  Directive as SvelteDirective,
  EachBlock as SvelteEachBlock,
  Element as SvelteElement,
  ElseBlock as SvelteElseBlock,
  IfBlock as SvelteIfBlock,
  MustacheTag as SvelteMustacheTag,
  Script as SvelteScript,
  ShorthandAttribute as SvelteShorthandAttribute,
  SpreadAttribute as SvelteSpreadAttribute,
  Style as SvelteStyle,
  StyleDirective as SvelteStyleDirective,
  TemplateNode as SvelteTemplateNode,
  Text as SvelteText
} from 'svelte/types/compiler/interfaces'

export { type SourceLocation } from '@vue-vapor/compiler-dom'
