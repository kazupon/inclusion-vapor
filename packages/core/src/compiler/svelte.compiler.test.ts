import { test, expect } from 'vitest'
import { compile, parse, preprocess } from 'svelte/compiler'

const svelteCode = `
<script>
  let count = 0
  const increment = () => {
    count += 1
  }
</script>

<button on:click={increment}>
  count is {count}
</button>
`

// NOTE: change to babel from acron in svelte, so not work compile (codegen)
test.skip('compile', () => {
  const returnValue = compile(svelteCode)
  // @ts-expect-error -- ignore
  delete returnValue['stats']
  expect(returnValue).toMatchSnapshot()
})

test('parser', () => {
  const returnValue = parse(svelteCode)
  expect(returnValue).toMatchSnapshot()
})

test('preprocess', async () => {
  const returnValue = await preprocess(svelteCode, {
    script(_params, ..._args) {
      // console.log('preprocess params', params)
      // console.log('preprocess args', args)
      return {
        code: `
import { onMount, computed, ref } from 'vue/vapor'
const count = ref(0)
onMount(() => {
  console.log('mounted')
})
const increment = computed(() => count.value + 1)
`
      }
    }
  })
  expect(returnValue).toMatchSnapshot()

  // const ast = parse(returnValue.code)
  // console.log('ast', (ast.instance?.content as unknown as File).program.body)
  // console.log('ast', ast.instance?.content)
})
