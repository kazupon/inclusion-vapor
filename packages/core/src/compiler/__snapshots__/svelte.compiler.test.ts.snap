// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parser 1`] = `
{
  "css": undefined,
  "html": {
    "children": [
      {
        "data": "
",
        "end": 1,
        "raw": "
",
        "start": 0,
        "type": "Text",
      },
      {
        "data": "

",
        "end": 84,
        "raw": "

",
        "start": 82,
        "type": "Text",
      },
      {
        "attributes": [
          {
            "end": 112,
            "expression": Node {
              "end": 111,
              "loc": SourceLocation {
                "end": Position {
                  "column": 27,
                  "line": 9,
                },
                "start": Position {
                  "column": 18,
                  "line": 9,
                },
              },
              "name": "increment",
              "start": 102,
              "type": "Identifier",
            },
            "modifiers": [],
            "name": "click",
            "start": 92,
            "type": "EventHandler",
          },
        ],
        "children": [
          {
            "data": "
  count is ",
            "end": 125,
            "raw": "
  count is ",
            "start": 113,
            "type": "Text",
          },
          {
            "end": 132,
            "expression": Node {
              "end": 131,
              "loc": SourceLocation {
                "end": Position {
                  "column": 17,
                  "line": 10,
                },
                "start": Position {
                  "column": 12,
                  "line": 10,
                },
              },
              "name": "count",
              "start": 126,
              "type": "Identifier",
            },
            "start": 125,
            "type": "MustacheTag",
          },
          {
            "data": "
",
            "end": 133,
            "raw": "
",
            "start": 132,
            "type": "Text",
          },
        ],
        "end": 142,
        "name": "button",
        "start": 84,
        "type": "Element",
      },
    ],
    "end": 142,
    "start": 1,
    "type": "Fragment",
  },
  "instance": {
    "attributes": [],
    "content": Node {
      "comments": [],
      "end": 73,
      "errors": [],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 7,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "program": Node {
        "body": [
          Node {
            "declarations": [
              Node {
                "end": 25,
                "id": Node {
                  "end": 21,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 11,
                      "line": 3,
                    },
                    "filename": undefined,
                    "identifierName": "count",
                    "start": Position {
                      "column": 6,
                      "line": 3,
                    },
                  },
                  "name": "count",
                  "start": 16,
                  "type": "Identifier",
                },
                "init": Node {
                  "end": 25,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 15,
                      "line": 3,
                    },
                    "filename": undefined,
                    "identifierName": undefined,
                    "start": Position {
                      "column": 14,
                      "line": 3,
                    },
                  },
                  "raw": "0",
                  "start": 24,
                  "type": "Literal",
                  "value": 0,
                },
                "loc": SourceLocation {
                  "end": Position {
                    "column": 15,
                    "line": 3,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 6,
                    "line": 3,
                  },
                },
                "start": 16,
                "type": "VariableDeclarator",
              },
            ],
            "end": 25,
            "kind": "let",
            "loc": SourceLocation {
              "end": Position {
                "column": 15,
                "line": 3,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 2,
                "line": 3,
              },
            },
            "start": 12,
            "type": "VariableDeclaration",
          },
          Node {
            "declarations": [
              Node {
                "end": 72,
                "id": Node {
                  "end": 43,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 17,
                      "line": 4,
                    },
                    "filename": undefined,
                    "identifierName": "increment",
                    "start": Position {
                      "column": 8,
                      "line": 4,
                    },
                  },
                  "name": "increment",
                  "start": 34,
                  "type": "Identifier",
                },
                "init": Node {
                  "async": false,
                  "body": Node {
                    "body": [
                      Node {
                        "end": 68,
                        "expression": Node {
                          "end": 68,
                          "left": Node {
                            "end": 63,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 9,
                                "line": 5,
                              },
                              "filename": undefined,
                              "identifierName": "count",
                              "start": Position {
                                "column": 4,
                                "line": 5,
                              },
                            },
                            "name": "count",
                            "start": 58,
                            "type": "Identifier",
                          },
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 14,
                              "line": 5,
                            },
                            "filename": undefined,
                            "identifierName": undefined,
                            "start": Position {
                              "column": 4,
                              "line": 5,
                            },
                          },
                          "operator": "+=",
                          "right": Node {
                            "end": 68,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 14,
                                "line": 5,
                              },
                              "filename": undefined,
                              "identifierName": undefined,
                              "start": Position {
                                "column": 13,
                                "line": 5,
                              },
                            },
                            "raw": "1",
                            "start": 67,
                            "type": "Literal",
                            "value": 1,
                          },
                          "start": 58,
                          "type": "AssignmentExpression",
                        },
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 14,
                            "line": 5,
                          },
                          "filename": undefined,
                          "identifierName": undefined,
                          "start": Position {
                            "column": 4,
                            "line": 5,
                          },
                        },
                        "start": 58,
                        "type": "ExpressionStatement",
                      },
                    ],
                    "end": 72,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 3,
                        "line": 6,
                      },
                      "filename": undefined,
                      "identifierName": undefined,
                      "start": Position {
                        "column": 26,
                        "line": 4,
                      },
                    },
                    "start": 52,
                    "type": "BlockStatement",
                  },
                  "end": 72,
                  "expression": false,
                  "generator": false,
                  "id": null,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 3,
                      "line": 6,
                    },
                    "filename": undefined,
                    "identifierName": undefined,
                    "start": Position {
                      "column": 20,
                      "line": 4,
                    },
                  },
                  "params": [],
                  "start": 46,
                  "type": "ArrowFunctionExpression",
                },
                "loc": SourceLocation {
                  "end": Position {
                    "column": 3,
                    "line": 6,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 8,
                    "line": 4,
                  },
                },
                "start": 34,
                "type": "VariableDeclarator",
              },
            ],
            "end": 72,
            "kind": "const",
            "loc": SourceLocation {
              "end": Position {
                "column": 3,
                "line": 6,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 2,
                "line": 4,
              },
            },
            "start": 28,
            "type": "VariableDeclaration",
          },
        ],
        "end": 73,
        "interpreter": null,
        "loc": SourceLocation {
          "end": Position {
            "column": 0,
            "line": 7,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "sourceType": "module",
        "start": 0,
        "type": "Program",
      },
      "start": 9,
      "type": "File",
    },
    "context": "default",
    "end": 82,
    "start": 1,
    "type": "Script",
  },
  "module": undefined,
}
`;

exports[`preprocess 1`] = `
{
  "code": "
<script>
import { onMount, computed, ref } from 'vue/vapor'
const count = ref(0)
onMount(() => {
  console.log('mounted')
})
const increment = computed(() => count.value + 1)
</script>

<button on:click={increment}>
  count is {count}
</button>
",
  "dependencies": [],
  "map": SourceMap {
    "ignoreList": [],
    "mappings": "",
    "names": [],
    "sourceRoot": undefined,
    "sources": [
      null,
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;
